compositor_render/src/transformations/web_renderer/renderer.rs

```diff
@@ -75,34 +90,188 @@ impl WebRenderer {
+        Self::copy_sources_to_buffers(wgpu_ctx, sources, &embedding_data.buffers);
+        self.copy_buffers_to_shared_memory(wgpu_ctx, sources, &embedding_data.buffers)?;
+
+        let resolutions = sources.iter().map(|texture| texture.resolution()).collect();
+        self.request_sender
+            .send(WebRendererThreadRequest::EmbedSources {
+                resolutions,
+                children_ids: embedding_data.children_ids.clone(),
+            })?;
+
+        Ok(())
+    }
+
+    fn copy_sources_to_buffers(
+        wgpu_ctx: &Arc<WgpuCtx>,
+        sources: &[&NodeTexture],
+        buffers: &[Arc<wgpu::Buffer>],
+    ) {
+        let mut encoder = wgpu_ctx.device.create_command_encoder(&Default::default());
+        for (texture, buffer) in sources.iter().zip(buffers) {
+            let Some(texture_state) = texture.state() else {
+                continue;
+            };
+            texture_state
+                .rgba_texture()
+                .copy_to_buffer(&mut encoder, buffer);
+        }
+
+        wgpu_ctx.queue.submit(Some(encoder.finish()));
+    }
+
+    fn copy_buffers_to_shared_memory(
+        &self,
+        wgpu_ctx: &Arc<WgpuCtx>,
+        sources: &[&NodeTexture],
+        buffers: &[Arc<wgpu::Buffer>],
+    ) -> Result<(), RenderWebsiteError> {
+        let mut pending_buffer_copies = Vec::new();
+        for (source_idx, (texture, buffer)) in sources.iter().zip(buffers).enumerate() {
+            let Some(texture_state) = texture.state() else {
+                continue;
+            };
+
+            let size = texture_state.rgba_texture().size();
+            let pending_copy = self.init_buffer_to_shmem_copy(source_idx, size, buffer.clone());
+            pending_buffer_copies.push(pending_copy);
+        }
+
+        wgpu_ctx.device.poll(wgpu::Maintain::Wait);
+
+        for pending in pending_buffer_copies {
+            pending()?;
+        }
+        Ok(())
+    }
+
+    fn init_buffer_to_shmem_copy(
+        &self,
+        source_idx: usize,
+        size: wgpu::Extent3d,
+        buffer: Arc<wgpu::Buffer>,
+    ) -> impl FnOnce() -> Result<(), RenderWebsiteError> + '_ {
+        let (s, r) = crossbeam_channel::bounded(1);
+        buffer
+            .slice(..)
+            .map_async(wgpu::MapMode::Read, move |result| {
+                if let Err(err) = s.send(result) {
+                    error!("channel send error: {err}")
+                }
+            });
+
+        move || {
+            r.recv().unwrap()?;
+
+            self.update_shared_memory(source_idx, buffer.clone(), size)?;
+            buffer.unmap();
+
+            Ok(())
+        }
+    }
+
+    pub fn frame_data(&self) -> Result<Option<Bytes>, RenderWebsiteError> {
+        let (response_sender, response_receiver) = new_response_channel();
+        self.request_sender
+            .send(WebRendererThreadRequest::GetRenderedWebsite { response_sender })?;
+
+        Ok(response_receiver.recv()?)
```

Comment:
```suggestion
        let result = self.request_sender.send(WebRendererThreadRequest::GetRenderedWebsite)?;
        Ok(result.wait()?)
```

It should look sth like this